jobbergate-agent/jobbergate_agent/utils/exception.py

```diff
@@ -0,0 +1,113 @@
+"""Core module for exception related operations"""
+import contextlib
+from asyncio import iscoroutinefunction
+from typing import Any, Callable, Iterable, Iterator, Mapping, Optional, Tuple, Type, Union
+
+from buzz import Buzz, get_traceback, reformat_exception
+from buzz.tools import DoExceptParams, noop
+
+
+class ClusterAgentError(Buzz):
+    """Raise exception when execution command returns an error"""
+
+
+class ProcessExecutionError(ClusterAgentError):
+    """Raise exception when execution command returns an error"""
+
+
+class AuthTokenError(ClusterAgentError):
+    """Raise exception when there are connection issues with the backend"""
+
+
+class SlurmrestdError(ClusterAgentError):
+    """Raise exception when slurmrestd raises any error"""
+
+
+class JobbergateApiError(ClusterAgentError):
+    """Raise exception when communication with Jobbergate API fails"""
+
+
+class JobSubmissionError(ClusterAgentError):
+    """Raise exception when a job cannot be submitted raises any error"""
+
+
+class SlurmParameterParserError(ClusterAgentError):
+    """Raise exception when Slurm mapper or SBATCH parser face any error"""
+
+
+@contextlib.asynccontextmanager
+async def handle_errors_async(
```

Comment:
I know I coded it, but nowadays I feel like it is too hacky, besides, mypy does not like it. I'll mark as technical dept.

Reply:
Tasks created: https://app.clickup.com/t/18022949/PENG-1660